File: app/src/main/java/com/anadol/mindpalace/view/adapters/MultipleFilter.java
package com.anadol.mindpalace.view.adapters;

import android.text.InputFilter;
import android.text.Spanned;

public class MultipleFilter implements InputFilter {

    @Override
    public CharSequence filter(CharSequence source, int start, int end, Spanned dest, int dstart, int dend) {
        //source - это новые символы, а dest - все остальные

        if (end - start > 1) {
            // Если одновременно было вставленно больше 1 символа
            return null;

        } else if (end - start == 1) {
            switch (source.charAt(start)) {

                case '\n':
                    if (dest.length() >= 1 && (dest.charAt(dend - 1) != ';')) return ";" + source;
                    if (dest.length() == 0) return "";
                    break;

                case ';':
                    if (dest.length() >= 1) {
                        return source + "\n";
                    } else {
                        return "";
                    }

            }
        }

        return null;
    }
}

File: app/src/main/java/com/anadol/mindpalace/view/adapters/MultipleFilter.kt
package com.anadol.mindpalace.view.adapters

import android.text.InputFilter
import android.text.Spanned

class MultipleFilter : InputFilter {

    override fun filter(
        source: CharSequence,
        start: Int,
        end: Int,
        dest: Spanned,
        dstart: Int,
        dend: Int
    ): CharSequence? {
        //source - это новые символы, а dest - все остальные
        if (end - start == 1) {

            when (source[start]) {
                '\n' -> {
                    if (dest.isNotEmpty() && dest[dend - 1] != ';') return ";$source"
                    if (dest.isEmpty()) return ""
                }
                ';' -> return if (dest.isNotEmpty()) "$source\n" else ""
            }
        }

        return null
    }
}
